README
Net-Shoutcast-Admin - Perl module to interact with Shoutcast servers

A quick module to interact with Shoutcast servers.  Currently supports fetching
information about the current status of a shoutcast server; future improvements
will add ability to perform actions such as kicking the current source, banning
IPs etc, and may add Icecast support too.

This module is still at a very early stage; if you find it useful, feel free
to let me know to help encourage further development :)


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install


DEPENDENCIES

LWP
XML::Simple


COPYRIGHT AND LICENCE

Copyright (C) 2008, David Precious

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

$Id$

Changes
Revision history for Net-Shoutcast-Admin

0.03  ????-??-??
       Some POD fixes

0.02  2008-03-19
       Improvements, some documentation fixes, adding new methods, handle
       a single listener properly by using XML::Simple's forceArray option.
       Still a fair bit to do; the N::S::A::Song/Listener classes are overly
       bloatly, I may re-implement them with Class::Accessor.
       
0.01  2008/02/12
       Initial release - beta code, still very limited, but seems to work
       so far :)



$Id$


eg/example.pl
#!/usr/bin/perl

use strict;
use warnings;

use lib '../lib';
use Net::Shoutcast::Admin;

# Simple usage example for Net::Shoutcast::Admin
# $Id$


my ($host, $port, $password) = @ARGV;

if (!$host || !$port || !$password) {
    die "Usage: $0 hostname port password";
}

my $shoutcast = Net::Shoutcast::Admin->new(
    host => $host,   port => $port,   admin_password => $password,
);

if ($shoutcast->source_connected) {
    my $listeners = $shoutcast->listeners;
    print "Stream is up, with $listeners listeners.\n";
    print "Current song is: " . $shoutcast->currentsong->title . "\n";
    
    for my $listener ( $shoutcast->listeners ) {
        printf "Listener from %s has been listening for %s\n",
            $listener->host, $listener->listen_time;
    }
    
    print "The last songs played are:\n";
    
    for my $song ($shoutcast->song_history) {
        print $song->title . "\n";
    }
    
} else {
    print "No source is currently connected.";
}

lib/Net/Shoutcast/Admin/Song.pm
package Net::Shoutcast::Admin::Song;
# $Id$

use warnings;
use strict;
use Carp;

use vars qw($VERSION);
$VERSION = '0.02';


=head1 NAME

Net::Shoutcast::Admin::Song - object to represent a song


=head1 DESCRIPTION

An object representing a song (either the current song, or an entry in the
song history).  Barely justifies being an object rather than a hashref
at the moment, but at some point I may be able to add other useful methods,
perhaps $song->lyrics to attempt to fetch lyrics.


=head1 SYNOPSIS

    use Net::Shoutcast::Admin;

    my $shoutcast = Net::Shoutcast::Admin->new(
                                    host => 'server hostname',
                                    port => 8000,
                                    admin_password => 'mypassword',
    );
    
    if ($shoutcast->source_connected) {
        my $song = $shoutcast->current_song;
        
        print "Current song is: " . $song->title;
    } else {
        print "No source is currently connected.";
    }
  
  
=head1 DESCRIPTION

Object representing a song, returned by Net::Shoutcast::Admin


=head1 INTERFACE 

=over 4

=item new

There's no reason to create instances of Net::Shoutcast::Admin::Song directly;
Net::Shoutcast::Admin creates and returns instances for you.

Having said that:

  $song = Net::Shoutcast::Admin::Song->new( %params );

Creates a new Net::Shoutcast::Admin object.  Takes a hash of options
as follows:

=over 4

=item I<title>

The title of the song

=item I<played_at>

The timestamp this song started playing.

=back

=cut

sub new {

    my ($class, %params) = @_;
    
    my $self = bless {}, $class;
        
    $self->{last_update} = 0;
    
    my %acceptable_params = map { $_ => 1 } 
        qw(title played_at);
    my %required_params = map { $_ => 1 } 
        qw(title);
    
    
    # make sure we haven't been given any bogus parameters:
    if (my @bad_params = grep { ! $acceptable_params{$_} } keys %params) {
        carp "Net::Shoutcast::Admin::Song does not recognise param(s) "
            . join ',', @bad_params;
        return;
    }
    
    $self->{$_} = $params{$_} for keys %acceptable_params;
    
    if (my @missing_params = grep { ! $self->{$_} } keys %required_params) {
        carp "Net::Shoutcast::Admin->new() must be supplied with params: "
            . join ',', @missing_params;
        return;
    }
    
    return $self;

}


=item title

Returns the artist

=cut

sub title {
    return shift->{title};
}

sub played_at {
    return shift->{played_at};
}



1; # Magic true value required at end of module
__END__


=back

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-net-shoutcast-admin@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

David Precious  C<< <davidp@preshweb.co.uk> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2008, David Precious C<< <davidp@preshweb.co.uk> >>. 
All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Net/Shoutcast/Admin/Listener.pm
package Net::Shoutcast::Admin::Listener;
# $Id$

use warnings;
use strict;
use Carp;

use vars qw($VERSION);
$VERSION = '0.01';


=head1 NAME

Net::Shoutcast::Admin::Listener - object to represent a listener


=head1 DESCRIPTION

An object representing a listener, returned by Net::Shoutcast::Admin.


=head1 SYNOPSIS

    use Net::Shoutcast::Admin;

    my $shoutcast = Net::Shoutcast::Admin->new(
                                    host => 'server hostname',
                                    port => 8000,
                                    admin_password => 'mypassword',
    );
    
    if ($shoutcast->source_connected) {
        my @listeners = $shoutcast->listeners;
        
        for my $listener (@listeners) {
            printf "Listener from %s, listening for %s",
                $listener->host, $listener->listen_time
            ;
        }
    } else {
        print "No source is currently connected.";
    }
  
  
=head1 DESCRIPTION

Object representing a listener, returned by Net::Shoutcast::Admin


=head1 INTERFACE 

=over 4

=item new

There's no reason to create instances of Net::Shoutcast::Admin::Listener 
directly; Net::Shoutcast::Admin creates and returns instances for you.

Having said that:

  $song = Net::Shoutcast::Admin::Listener->new( %params );

Creates a new Net::Shoutcast::Admin::Listener object.  Takes a hash of options
as follows:

=over 4

=item I<host>

The host from which this listener is connected

=item I<connect_time>

The number of seconds this listener has been connected

=item I<underruns>

The number of buffer underruns this listener has suffered

=item I<agent>

The software this user is reportedly using to listen

=back

=cut

sub new {

    my ($class, %params) = @_;
    my $self = bless {}, $class;
        
    $self->{last_update} = 0;
    
    my %acceptable_params = map { $_ => 1 } 
        qw(host connect_time underruns agent);
    
    # make sure we haven't been given any bogus parameters:
    if (my @bad_params = grep { ! $acceptable_params{$_} } keys %params) {
        carp "Net::Shoutcast::Admin::Listener does not recognise param(s) "
            . join ',', @bad_params;
        return;
    }
    
    $self->{$_} = $params{$_} for keys %acceptable_params;
    
    if (my @missing_params = grep { ! $self->{$_} } keys %acceptable_params) {
        carp "Net::Shoutcast::Admin::Listener->new() must be supplied with "
        . "params: "
            . join ',', @missing_params;
        return;
    }
    
    return $self;

}


=item host

Returns the hostname this listener is connected from

=cut

sub host { return shift->{host} }


=item listen_time

Returns the number of seconds this listener has been connected

=cut

sub listen_time { return shift->{connect_time} }



=item underruns

Returns the number of buffer underruns this listener has suffered

=cut

sub underruns { return shift->{underruns} }


=item agent

Returns the agent this listener is reportedly using

=cut

sub agent { return shift->{agent} }


1; # Magic true value required at end of module
__END__


=back

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-net-shoutcast-admin@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

David Precious  C<< <davidp@preshweb.co.uk> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2008, David Precious C<< <davidp@preshweb.co.uk> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Net/Shoutcast/Admin.pm
package Net::Shoutcast::Admin;
# $Id$

use warnings;
use strict;
use Carp;
use Net::Shoutcast::Admin::Song;
use Net::Shoutcast::Admin::Listener;
use URI::Escape;
use LWP::UserAgent;
use XML::Simple;

use vars qw($VERSION);
$VERSION = '0.02';


=head1 NAME

Net::Shoutcast::Admin - administration of Shoutcast servers


=head1 VERSION

This document describes Net::Shoutcast::Admin version 0.0.2


=head1 SYNOPSIS

    use Net::Shoutcast::Admin;

    my $shoutcast = Net::Shoutcast::Admin->new(
                                    host => 'server hostname',
                                    port => 8000,
                                    admin_password => 'mypassword',
    );
    
    if ($shoutcast->source_connected) {
        printf "%s is currently playing %s by %s",
            $shoutcast->dj_name,
            $shoutcast->currentsong->title,
            $shoutcast->currentsong->artist
        ;
    } else {
        print "No source is currently connected.";
    }
  
  
=head1 DESCRIPTION

A module to interact with Shoutcast servers to retrieve information about
their current status (and perhaps in later versions of the module, to also
control the server in various ways).


=head1 INTERFACE 

=over 4

=item new

$shoutcast = Net::Shoutcast::Admin->new( %params );

Creates a new Net::Shoutcast::Admin object.  Takes a hash of options
as follows:

=over 4

=item host

The hostname of the Shoutcast server you wish to query.

=item port

The port on which Shoutcast is running.  Defaults to 8000 if not specified.

=item admin_password

The admin password for the Shoutcast server.

=item timeout

The number of seconds to wait for a response.  Defaults to 10 seconds if
not specified.

=item agent

The HTTP User-Agent header which will be sent in HTTP requests to the Shoutcast
server.  If not supplied, a suitable default will be used.

=back

=cut

sub new {

    my ($class, %params) = @_;
    
    my $self = bless {}, $class;
        
    $self->{last_update} = 0;
    
    my %acceptable_params = map { $_ => 1 } 
        qw(host port admin_password timeout agent);
    
    # make sure we haven't been given any bogus parameters:
    if (my @bad_params = grep { ! $acceptable_params{$_} } keys %params) {
        carp "Net::Shoutcast::Admin does not recognise param(s) "
            . join ',', @bad_params;
        return;
    }
    
    # 
    $self->{$_} = $params{$_} for keys %acceptable_params;
    
    # set decent defaults for optional params:
    $self->{port}    ||= 8000;
    $self->{timeout} ||= 10;
    # In my initial testing, it seems the Shoutcast server will not respond
    # with the status XML if the User-Agent does not contain "Mozilla"
    # (Why?  That's just lame!)
    $self->{agent}   ||= "Mozilla";
    
    if (my @missing_params = grep { ! $self->{$_} } keys %acceptable_params) {
        carp "Net::Shoutcast::Admin->new() must be supplied with params: "
            . join ',', @missing_params;
        return;
    }
    
    # get an LWP::UserAgent object to make our requests with:
    my $ua = new LWP::UserAgent;
    if ($ua) {
        $ua->agent(   $self->{agent}   );
        $ua->timeout( $self->{timeout} );
        $self->{ua} = $ua;
    } else {
        warn "Failed to create LWP::UserAgent object";
        return;
    }
    
    return $self;

}


# if we haven't fetched the status XML recently, fetch it.  Returns true if
# we either fetched it successfully or it was fresh enough to not need
# re-fetching, or false if we couldn't get it.
sub _update_if_necessary {
    my $self = shift;
    if ($self->{last_update} and (time - $self->{last_update}) < 5) {
            # status was updated not long ago
            return 1;
    }
    
    my ($fetched, $msg) = $self->_fetch_status_xml;
    if (!$fetched) {
        warn "Failed to fetch status from Shoutcast server: $msg";
        return;
    }
    
    # all good.
    return 1;
}


sub _fetch_status_xml {
    my $self = shift;
        
    my ($host, $port) = @$self{qw(host port)};
    my $pass = URI::Escape::uri_escape( $self->{admin_password} );
    
    my $url = "http://$host:$port/admin.cgi?pass=$pass&mode=viewxml";
    
    my $response = $self->{ua}->get($url);
    
    if (!$response->is_success) {
        my $err = "Failed to fetch status XML - " . $response->status_line;
        carp $err;
        return wantarray? (0, $err) : 0;
    }
    
    my $data = XML::Simple::XMLin($response->content, 
        forceArray => [qw(LISTENER SONG)]);
    
    if (!$data) {
        return wantarray? (0, 'Failed to parse XML') : 0;
    }
    
    $self->{data} = $data;
    $self->{last_update} = time;
    
    return wantarray? (1, undef) : 1;
}



=item currentsong

Returns a <Net::Shoutcast::Admin::Song> object representing the current
song.

=cut

sub currentsong {
    my $self = shift;
    $self->_update_if_necessary or return;
    
    my $song = Net::Shoutcast::Admin::Song->new( 
        title => $self->{data}->{SONGTITLE}
    );   
    return $song;
}


=item song_history

Returns a list of L<Net::Shoutcast::Admin::Song> objects representing
the the last few songs played

=cut

sub song_history {
    my $self = shift;
    my @song_objects;
    $self->_fetch_status_xml;
    
    for my $song (@{ $self->{data}->{SONGHISTORY}->{SONG} }) {
        push @song_objects, Net::Shoutcast::Admin::Song->new(
            title => $song->{TITLE},
            played_at => $song->{PLAYEDAT},
        );
    }
    
    return (@song_objects);
}


=item listeners

In scalar context, returns the number of listeners currently connected.
In list context, returns a list of L<Net::Shoutcast::Admin::Listener>
objects representing each listener.

=cut

sub listeners {
    my $self = shift;
    $self->_fetch_status_xml;
    
    if (!wantarray) {
        # okay, it's nice and simple:
        return $self->{data}->{CURRENTLISTENERS};
    } else {
        # okay, we need to return a list of N:S:A::Listener objects:
        return if !$self->{data}->{CURRENTLISTENERS};
        
        my @listener_objects;
        for my $listener (@{ $self->{data}->{LISTENERS}->{LISTENER} }) {
            push @listener_objects, Net::Shoutcast::Admin::Listener->new(
                host         => $listener->{HOSTNAME},
                connect_time => $listener->{CONNECTTIME},
                underruns    => $listener->{UNDERRUNS} || 5,
                agent        => $listener->{USERAGENT},
            );
        }
            
        return (@listener_objects);
    }
}


=item source_connected

Returns true if the stream is currently up (a source is connected and streaming
audio to the server)

=cut

sub source_connected {
    my $self = shift;
    $self->_fetch_status_xml;   
    return ($self->{data}->{STREAMSTATUS});
}


1; # Magic true value required at end of module
__END__



=back

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-net-shoutcast-admin@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

David Precious  C<< <davidp@preshweb.co.uk> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2008, David Precious C<< <davidp@preshweb.co.uk> >>. 
All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
